### PR TITLE
merge from multinode to master

### PR DESCRIPTION
**Pull request by josephdviviano**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
This contains a large number of changes from the multinode branch, to be pushed into master. Future multinode development will occur as branches off of master.

This contains:

- A multinode-compatible replay buffer.
- Changes to probability calculations, which now occur in log-prob space for numerical stability.
- An improved selection of off-policy sampling methods.
- Changes to the hypergrid scripts to support new multinode training features.

### CODE REVIEW

File: src/gfn/estimators.py [hunk 1]
```diff
@@ -561,124 +706,82 @@ def to_probability_distribution(
...
-
-        probs = {}
+        logits[Ga.EDGE_CLASS_KEY][no_possible_edge_class] = 0.0
+
+        # Check if no possible node can be added; if either class OR index has no
+        # valid options, disallow ADD_NODE in action type and set harmless finite
+        # fallbacks for the unused components.
+        no_possible_node_class = torch.isneginf(logits[Ga.NODE_CLASS_KEY]).all(-1)
+        no_possible_node_index = torch.isneginf(logits[Ga.NODE_INDEX_KEY]).all(-1)
+        no_possible_add_node = no_possible_node_class | no_possible_node_index
```
Comment (younik) [inline src/gfn/estimators.py:743]: we really need to fix this first, because I think it would make the add action impossible (like triangle tutorial giving 0%)

Or we temporarly revert to the previous workflow that converts to probability

File: src/gfn/utils/distributions.py [hunk 2]
```diff
@@ -143,18 +165,34 @@ def log_prob(self, sample: torch.Tensor) -> torch.Tensor:
...
             if self.is_backward:
                 log_prob_node_index_all = self.dists[
                     GraphActions.NODE_INDEX_KEY
                 ].log_prob(
                     sample[..., GraphActions.ACTION_INDICES[GraphActions.NODE_INDEX_KEY]]
                 )
+                assert torch.isfinite(
+                    log_prob_node_index_all[add_node_idx]
+                ).all(), (
+                    "add_node_idx is indexing masked values in log_prob_node_index_all"
```
[the remark this answers is not in the document]
Comment (younik) [reply src/gfn/utils/distributions.py:187]: yes, I have seen it, but I am still confused why it happen

We have -inf logprobs, because the logits are -inf, but this should entails that sampling the action is impossible

[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/utils/distributions.py:191]: I think the issue is we never actually call the backward branch of the GraphActionDistribution.sample()

File: src/gfn/samplers.py [hunk 3]
```diff
@@ -100,6 +100,30 @@ def sample_actions(
...
+                # Compute next index per graph from current states
+                # states is already masked to active rows for this call
+                next_indices = []
+                for graph in states.data.flat:  # type: ignore[attr-defined]
+                    n = 0 if graph.x is None else graph.x.size(0)
+                    next_indices.append(n)
+                next_indices_tensor = torch.as_tensor(
+                    next_indices, device=actions.device, dtype=actions.tensor.dtype
+                )
+                actions.node_index[add_node_mask] = next_indices_tensor[add_node_mask]
```
Comment (josephdviviano) [inline src/gfn/samplers.py:125]: So I really don't love having this logic here, so I'm open to suggestions, but this seems to fix the issue.

**Comment by josephdviviano** [comment]
@hyeok9855 can you have a quick look at @younik's recent commit https://github.com/GFNOrg/torchgfn/pull/395/commits/6511290b86c4663ea2908907005303086dfda945

which resolves the graph sampling error.

File: src/gfn/states.py [hunk 4]
```diff
@@ -870,14 +870,19 @@ def forward_masks(self) -> TensorDict:
         node_class_masks = torch.ones(
             self.data.size, self.num_node_classes, dtype=torch.bool, device=self.device
         )
-
+        node_index_masks = torch.zeros(
+            self.data.size, max_nodes + 1, dtype=torch.bool, device=self.device
+        )
```
Comment (hyeok9855) [inline src/gfn/states.py:875]: Why `max_nodes + 1` while in the backward_masks it's `max_nodes`?

**Comment by josephdviviano** [comment]
derpy git commit message
